network/endpoint_linux.go

```diff
@@ -53,205 +54,200 @@ func (nw *network) newEndpointImpl(
+						nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+					}
+
+					epClient = NewOVSEndpointClient(
+						nw,
+						epInfo,
+						hostIfName,
+						contIfName,
+						vlanid,
+						localIP,
+						nl,
+						ovsctl.NewOvsctl(),
+						plc)
 				}
-				epClient = NewTransparentVlanEndpointClient(nw, epInfo, hostIfName, contIfName, vlanid, localIP, nl, plc)
+			} else if nw.Mode != opModeTransparent {
+				logger.Info("Bridge client")
+				epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
+			} else if epInfo.NICType == cns.DelegatedVMNIC {
+				logger.Info("Secondary client")
+				epClient = NewSecondaryEndpointClient(nl, plc, ep)
 			} else {
-				logger.Info("OVS client")
-				if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
-					nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
-				}
-
-				epClient = NewOVSEndpointClient(
-					nw,
-					epInfo,
-					hostIfName,
-					contIfName,
-					vlanid,
-					localIP,
-					nl,
-					ovsctl.NewOvsctl(),
-					plc)
+				logger.Info("Transparent client")
+				epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 			}
-		} else if nw.Mode != opModeTransparent {
-			logger.Info("Bridge client")
-			epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
-		} else {
-			logger.Info("Transparent client")
-			epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 		}
-	}
 
-	// Cleanup on failure.
-	defer func() {
-		if err != nil {
-			logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
-			endpt := &endpoint{
-				Id:                       epInfo.Id,
-				IfName:                   contIfName,
-				HostIfName:               hostIfName,
-				LocalIP:                  localIP,
-				IPAddresses:              epInfo.IPAddresses,
-				DNS:                      epInfo.DNS,
-				VlanID:                   vlanid,
-				EnableSnatOnHost:         epInfo.EnableSnatOnHost,
-				EnableMultitenancy:       epInfo.EnableMultiTenancy,
-				AllowInboundFromHostToNC: epInfo.AllowInboundFromHostToNC,
-				AllowInboundFromNCToHost: epInfo.AllowInboundFromNCToHost,
+		//nolint:gocritic
+		defer func(client EndpointClient, contIfName string) {
+			// Cleanup on failure.
+			if err != nil {
+				logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
+				if containerIf != nil {
+					client.DeleteEndpointRules(ep)
+				}
+				// set deleteHostVeth to true to cleanup host veth interface if created
+				//nolint:errcheck // ignore error
+				client.DeleteEndpoints(ep)
 			}
+		}(epClient, contIfName)
 
-			if containerIf != nil {
-				endpt.MacAddress = containerIf.HardwareAddr
-				epClient.DeleteEndpointRules(endpt)
+		//nolint:wrapcheck // ignore wrap check
+		err = func() error {
```

Comment:
why do you introduce this wrapper?

Reply:
we want to cleanup on error per epInfo using it's respective epClient

Reply:
why we do we need wrapper specifically? why just returning error wouldn't work? defer will be called anyways

Reply:
the defer to close namespace won't happen (for non error scenario) before the next loop if it's not wrapped

Reply:
added comment